### ISSUE TITLE
Bastion of Remembrance/Molten Echoes - CME with odd rollback behavior

### ISSUE THREAD

**Issue by rystan** [magefree/mage#7565]
```
Game exception occurred: java.util.ConcurrentModificationException
Server version: 1.4.48V1 (build: 2021-02-13 16:43)
java.util.ArrayList$Itr.checkForComodification(ArrayList.java:909)
java.util.ArrayList$Itr.next(ArrayList.java:859)
mage.abilities.AbilitiesImpl.<init>(AbilitiesImpl.java:35)
mage.abilities.AbilitiesImpl.copy(AbilitiesImpl.java:42)
mage.abilities.AbilitiesImpl.copy(AbilitiesImpl.java:21)
mage.MageObjectImpl.<init>(MageObjectImpl.java:75)
mage.cards.CardImpl.<init>(CardImpl.java:114)
mage.game.permanent.PermanentImpl.<init>(PermanentImpl.java:123)
mage.game.permanent.PermanentToken.<init>(PermanentToken.java:39)
mage.game.permanent.PermanentToken.copy(PermanentToken.java:102)
mage.game.permanent.PermanentToken.copy(PermanentToken.java:17)
mage.game.GameImpl.getLastKnownInformation(GameImpl.java:2923)
mage.game.GameImpl.getObject(GameImpl.java:377)
mage.abilities.AbilityImpl.hasSourceObjectAbility(AbilityImpl.java:984)
mage.abilities.AbilityImpl.isInUseableZone(AbilityImpl.java:941)
mage.abilities.StaticAbility.isInUseableZone(StaticAbility.java:36)
mage.abilities.effects.ContinuousEffects.getLayeredEffects(ContinuousEffects.java:186)
mage.abilities.effects.ContinuousEffects.getLayeredEffects(ContinuousEffects.java:163)
mage.abilities.effects.ContinuousEffects.apply(ContinuousEffects.java:968)
mage.game.GameState.applyEffects(GameState.java:602)
mage.game.GameImpl.applyEffects(GameImpl.java:1576)
mage.game.GameImpl.playPriority(GameImpl.java:1377)
mage.game.turn.Step.priority(Step.java:61)
mage.game.turn.Phase.playStep(Phase.java:184)
mage.game.turn.Phase.play(Phase.java:89)
mage.game.turn.Turn.play(Turn.java:128)
mage.game.GameImpl.playTurn(GameImpl.java:911)
mage.game.GameImpl.play(GameImpl.java:821)
mage.game.GameImpl.start(GameImpl.java:797)
mage.server.game.GameWorker.call(GameWorker.java:34)
java.util.concurrent.FutureTask.run(FutureTask.java:266)
java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
java.lang.Thread.run(Thread.java:748)
```

Had [[Molten Echoes]] (targeting 'Human' creature type).  Also had [[Bastion of Remembrance]] in play.  Cast [[Adriana, Captain of the Guard]] which is `Legendary Human` which triggers Molten Echoes to make a copy of it, which due to Legendary rule forced me to pick a copy to keep.  Selected to keep the non-token one, which triggers the Bastion of Remembrance ability onto the stack and that seemed to be when the CME error occurs and rolls back the Echoes, and Adriana to my hand along with any land I play in the same main phase.  The rollback doesn't appear to go back start of turn, but triggers the end of the current main phase and was repeatable error in both main phases.  The first time after I clicked Okay on the error dialog it immediately went to "beginning of combat" and during the second main it triggered "end step".

**Comment by JayDi85** [comment]
Related to #7170

**Comment by JayDi85** [comment]
@rystan is that AI or human game?

**Comment by rystan** [comment]
Playing against AI.

**Comment by rystan** [comment]
I still have the game in progress, if you want/need additional information at this point.

**Comment by JayDi85** [comment]
Nope, it useless (game data can be broken already). There are must be steps to reproduce from the start. That bug can be related to another cards or effects that broke the game state data, or it can be related to wrong targeting in AI games (example: AI gives gained ability to source card).

**Comment by JayDi85** [comment]
Can be fixed in 1664ee01cf30af2e85d725d63f22808ffdf2d61f. If not then re-open it.
